### ISSUE TITLE
[bug]: chan_iax2: Packet decryption failures when dialing quickly

### ISSUE THREAD

**Issue by InterLinked1** [asterisk/asterisk#457]
### Severity

Major

### Versions

21.0.0

### Components/Modules

chan_iax2

### Operating Environment

Debian 12

### Frequency of Occurrence

Constant

### Issue Description

When dialing digits quickly during call setup, IAX2 seems to have some kind of race condition where it will cause packet decryption to fail (this is with encryption enabled for the peer). The log is spammed thusly:

```
[2023-11-23 02:31:37] NOTICE[116137]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116132]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116133]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116139]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116135]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116137]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116132]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116133]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116139]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116135]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116137]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116132]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116133]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116139]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116135]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116137]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116132]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116133]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
[2023-11-23 02:31:37] NOTICE[116139]: chan_iax2.c:10252 socket_process_helper: Packet Decrypt Failed!
```

The originating side ends up getting stuff like this:
```
[2023-11-22 21:31:39.367] WARNING[1007420]: chan_iax2.c:1213 iax_error_output: Information element length exceeds message size
[2023-11-22 21:31:39.367] WARNING[1007420]: chan_iax2.c:10290 socket_process_helper: Undecodable frame received from 'REDACTED:4569'
```

No audio works for the call, the entire call is borked and this goes on until manual clearing.

This is the basic setup:

```
exten => 8,1,Dial(IAX2/tieline:password@main.example.com/s,,g) ; tie trunk to the main office
	same => n,Hangup()
```

On the originating side in iax.conf:
```
[tieline]
type = friend
username = tieline
context = from-tieline
```

On the remote side:

```
[tieline]
type=friend
username = tieline
context=tieline-in-001
auth=md5
secret=password
forceencryption=yes
requirecalltoken=yes
host = dynamic
```

To reproduce the issue, using any Hayes-compatible modem, on the originating Asterisk system, do:

```
ATS11=100 ; Set the ATS 11 register to a relatively small value (but this is higher than the default, I believe). Asterisk DTMF detection is quite poor so modems don't work without manually increasing the spacing between digits when dialing.
ATDT881
```

At this point, the issue above is reproduced. The use of a modem is not relevant in any way other than modems by default tend to dial extremely quickly, which is what exposed this issue in the first place.

However, if doing `ATDT8,81` instead of `ATDT881`, interjecting a pause after dialing the local code assigned to this trunk, the issue does not happen. Without looking into this further, my guess is that if a DTMF digit is sent very quickly after the trunk is set up, this happens before something that should happen at the beginning of call setup has happened (perhaps a DTMF frame being sent before an encryption setup frame or the like?) and this screws up the entire call.

This problem could affect other types of frames as well, so probably isn't DTMF-specific inherently, but DTMF is something that end users can control and thus easily manipulate to trigger this issue.

This is less likely to happen with `ATS11=250` consistently, but still does happen occasionally. 250 is the maximum value for the ATS11 register so there is literally no way to make the modem dial any slower. However, this range of timing suggests that a window as long as 250 ms perhaps can exist where this bug can be triggered and anyone (or anything) dialing faster than that will trigger this bug.

Manually inserting a pause `ATDT8,81` ensures that the call has time to set up appropriately before sending the next DTMF frame. The only other workaround at this time is to avoid using DTMF altogether and use dial pulse, e.g. `ATDP881`, since this inherently causes the first DTMF frame sent over the trunk to get delayed.

### Asterisk Issue Guidelines

- [X] Yes, I have read the Asterisk Issue Guidelines

**Comment by seanbright** [comment]
Packet capture? `iax2 set debug on` output?

**Comment by InterLinked1** [comment]
> Packet capture? `iax2 set debug on` output?

Yeah, I don't know what I was thinking. It was late last night and I didn't think to do so.

I've tried reproducing this morning and though it was consistent last night, having turned the modems off overnight, I can't seem to quite reproduce it this morning, so there may have been other contributing environmental factors. I'm getting occasional digit loss as usual at higher speeds but it's not triggering that particular issue. I also restarted Asterisk on the home system for an unrelated reason. If I'm able to reproduce again, I'll update this issue with debug on from both sides.
